daemon/network.go

```diff
@@ -870,7 +870,8 @@ func buildCreateEndpointOptions(c *container.Container, n *libnetwork.Network, e
 	// to which container was connected to on docker run.
 	// Ideally all these network-specific endpoint configurations must be moved under
 	// container.NetworkSettings.Networks[n.Name()]
-	if nwName == c.HostConfig.NetworkMode.NetworkName() || (nwName == defaultNetName && c.HostConfig.NetworkMode.IsDefault()) {
+	netMode := c.HostConfig.NetworkMode
+	if nwName == netMode.NetworkName() || strings.HasPrefix(n.ID(), netMode.NetworkName()) || (nwName == defaultNetName && netMode.IsDefault()) {
```

Comment:
Of course, this is still wrong... `netMode.NetworkName()` might be a partial prefix matching multiple things, for instance a network ID and a network name.